README.md
# echoserver

```
                       _/
    _/_/      _/_/_/  _/_/_/      _/_/      _/_/_/    _/_/    _/  _/_/  _/      _/    _/_/    _/  _/_/
 _/_/_/_/  _/        _/    _/  _/    _/  _/_/      _/_/_/_/  _/_/      _/      _/  _/_/_/_/  _/_/
_/        _/        _/    _/  _/    _/      _/_/  _/        _/          _/  _/    _/        _/
 _/_/_/    _/_/_/  _/    _/    _/_/    _/_/_/      _/_/_/  _/            _/        _/_/_/  _/
 ```

echoserver is an HTTP server implementation that is useful for testing out provisioning
infrastructure or for testing application integration with a third-party API.

The implementation is created in Go utilizing [Cobra](https://github.com/spf13/cobra) 
and the [gorilla/mux](https://github.com/gorilla/mux) libraries.

## Getting started

## make commands

```
Usage:
  make <target>

Targets:
  uninstall            Delete all of the plugins/tools versions of the asdf apps. Mainly for testing if install works.
  install              Install all plugins/tool versions of asdf apps.
  begin                Install all dependencies and create the cluster, installing all of the infrastructure.
  create               Creates the cluster and adds the baseline resources.
  kustomize            Install Kustomize files.
  destroy              WARNING: Destroys the cluster entirely!
  start                Starts the cluster from suspension. The cluster must be created first before this can work.
  stop                 Stops the cluster, putting into suspension. The cluster must have been created and running for this to work.
  up                   Start the Tilt server.
  exec                 Exec into a pod in the cluster for testing purposes.
  watch                Run gow to restart the server on file saves.
  build                Build the server for the current architecture
  build-linux          Build the server for the linux architecture
  build-docker         Build a Docker image of the server
  run                  Run the application in a container
  help                 Show help
```

**Dependencies**:
- [go 1.17](https://go.dev/) (Required)

You can start the project by running the server:
```
go run main.go
```

Then in a browser window or in another shell you can execute commands to the server:
```
$ curl http://localhost:8080/200
```

### Configuration
```
An HTTP echo server implementation in Go. This will echo back to you the status
code and information you send it. This is useful for testing infrastructure.

Usage:
  echoserver [flags]
  echoserver [command]

Available Commands:
  completion  Generate the autocompletion script for the specified shell
  help        Help about any command
  live        Check for liveness file.

Flags:
      --banner                 Print the banner on server start. (default true)
      --config string          config file (default is $HOME/.echo.yaml)
  -h, --help                   help for echoserver
      --host string            The host to run the server on. (default "localhost")
      --idle-timeout int       Idle timeout in seconds. (default 60)
      --liveness-path string   Path to put the liveness file. (default "/tmp/echoserver-live")
      --log-level string       Logging level. Valid choices are 'info', 'debug', and 'trace'. (default "info")
      --metrics-port int       The metrics port to run the server on. This can be the same as the host port, but is set to a separate port by default for security. (default 9000)
      --port int               The port to run the server on. (default 8080)
      --read-timeout int       Read timeout in seconds. (default 15)
      --wait int               Wait timeout in seconds. (default 15)
      --write-timeout int      Write timeout in seconds. (default 15)

Use "echoserver [command] --help" for more information about a command.
```

## Local Kubernetes Development Example

This project comes with an example deployment into a local Kubernetes cluster using `k3d`.

**Dependencies**:
- [make](https://gnu.org/software/make/) (Not required, but encouraged for convenience of commands)
- [Docker](https://https://www.docker.com/) (Required for local Kubernetes development/examples)
- [asdf-vm](https://asdf-vm.com/) (Not required, but encouraged to manage dependencies)
- [k3d](https://k3d.io/) (Required only for local Kubernetes development/examples)
- [Tilt](https://tilt.dev/) (Convenient for local Kubernetes development)
- [kubectl](https://kubernetes.io/docs/tasks/tools/) (Convenient for local Kubernetes development)

This project utilizes the excellent [asdf version manager](https://github.com/asdf-vm/asdf) to 
manage all dependencies outside of Docker and `make`. `asdf` is not required, but it makes the
installation easier.

To get started, you can 
[install asdf](http://asdf-vm.com/guide/getting-started.html#getting-started) and utilize the 
provided `make install` command, or you can install the versions of the applications in the 
[.tool-versions](./.tool-versions) file directly.

From there, the rest of the `make` commands should get you started.

### Default configuration

The [k3d config file](./k3d-config.yaml) file has all of the default configuration for the project.
By default, ports `30080-30099` will be allocated on the host machine for services running in the
cluster. If any of these ports are already taken by another process on the host, the creation of
the cluster will fail.

The cluster is started with 1 server (control plane node) and 1 agent (worker node). Traefik is
disabled due to a limitation on how services are exposed in k3d. Allocating the default set of 
ports and using `NodePort` to set up the services made the process a lot easier.

### kube-ops-view

The [kube-ops-view](https://codeberg.org/hjacobs/kube-ops-view) project is installed into the
cluster when running the `begin` or `kustomize` commands. This will take the Kustomization files
from the `hjacobs` repo and install kube-ops-view directly into the cluster. Kube-ops-view can
then be accessed by going to [http://localhost:30099](http://localhost:30099) in the browser.

![kube-ops-view](./docs/readme/kube-ops-view.png)

### exec-pod

For convenience, an execution pod is installed into the cluster. This allows you to run commands
and install software inside of an [alpine](https://alpinelinux.org/) container to run commands on
the cluster, if you wish. This is useful for mocking out a service that is running inside of the
cluster.

```
$ make exec
```
Which is the same as:
```
$ kubectl --namespace default exec --stdin=true --tty=true pod/exec-pod -- sh
```

![make exec](./docs/readme/make-exec.svg)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] - 2022-03-22
Initial release


internal/routes.go
package internal

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"time"

	"github.com/gorilla/mux"
	"github.com/rs/zerolog/log"
	"gopkg.in/yaml.v2"
)

type Status struct {
	Code int    `json:"code" yaml:"code"`
	Text string `json:"text" yaml:"text"`
}

func getStatus(r *http.Request) (response Status, err error) {
	vars := mux.Vars(r)
	code := vars["code"]
	status := Status{}

	status.Code, err = strconv.Atoi(code)
	if err != nil {
		log.Error().Msg("Error converting http status code string to integer")
		return status, errors.New("String to int conversion error")
	}
	status.Text = http.StatusText(status.Code)

	return status, nil
}

func handlePlaintext(w http.ResponseWriter, r *http.Request) {
	status, _ := getStatus(r)
	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
	w.WriteHeader(status.Code)
	io.WriteString(w, fmt.Sprintf("%d %s", status.Code, status.Text))
}

func handleJson(w http.ResponseWriter, r *http.Request) {
	status, _ := getStatus(r)
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status.Code)
	json.NewEncoder(w).Encode(status)
}

func handleYaml(w http.ResponseWriter, r *http.Request) {
	status, _ := getStatus(r)
	w.Header().Set("Content-Type", "application/yaml")
	w.WriteHeader(status.Code)
	yaml.NewEncoder(w).Encode(status)
}

func handleHealthcheck(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(http.StatusOK)
}

func handlePing(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(http.StatusOK)
	io.WriteString(w, fmt.Sprintf("pong"))
}

func handleSync(w http.ResponseWriter, r *http.Request) {
	time.Sleep(time.Second * 5)
	w.WriteHeader(http.StatusOK)
	io.WriteString(w, fmt.Sprintf("completed"))
}

func handleSyncSleep(w http.ResponseWriter, r *http.Request) {
	sleepTime := mux.Vars(r)["sleep"]
	sleep, err := strconv.Atoi(sleepTime)
	log.Debug().Msg(sleepTime)
	if err != nil {
		w.WriteHeader(600)
		io.WriteString(w, fmt.Sprintf("echoserver error: %v", err))
	} else {
		time.Sleep(time.Second * time.Duration(sleep))
		w.WriteHeader(http.StatusOK)
		io.WriteString(w, fmt.Sprintf("completed"))
	}
}

func handleFavicon(w http.ResponseWriter, r *http.Request) {
	w.WriteHeader(http.StatusNotFound)
}

func AddRoutes(router *mux.Router) {
	router.HandleFunc("/health", handleHealthcheck)
	router.HandleFunc("/ping", handlePing)

	router.HandleFunc("/favicon.ico", handleFavicon)

	router.HandleFunc("/sync", handleSync)
	router.HandleFunc("/sync", handleSyncSleep).Queries("sleep")

	router.HandleFunc("/{code:[0-9]+}", handlePlaintext).Queries("output", "text")
	router.HandleFunc("/{code:[0-9]+}", handleJson).Queries("output", "json")
	router.HandleFunc("/{code:[0-9]+}", handleYaml).Queries("output", "yaml")

	router.HandleFunc("/{code:[0-9]+}", handlePlaintext).Headers("echoserver-output", "text")
	router.HandleFunc("/{code:[0-9]+}", handleJson).Headers("echoserver-output", "json")
	router.HandleFunc("/{code:[0-9]+}", handleYaml).Headers("echoserver-output", "yaml")

	router.HandleFunc("/text/{code:[0-9]+}", handlePlaintext)
	router.HandleFunc("/json/{code:[0-9]+}", handleJson)
	router.HandleFunc("/yaml/{code:[0-9]+}", handleYaml)

	router.HandleFunc("/{code:[0-9]+}", handlePlaintext)
}


internal/server.go
package internal

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/gorilla/mux"
	"github.com/rs/zerolog/log"
)

type httpServer interface {
	Shutdown(ctx context.Context) error
	ListenAndServe() error
}

type Server struct {
	http.Server
	Name string
}

type Options struct {
	LivenessFilePath string
	Host             string
	Port             int
	MetricsPort      int
	Wait             int
	WriteTimeout     int
	ReadTimeout      int
	IdleTimeout      int
}

func loggingMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		log.Debug().Msgf("Started request to %s", r.RequestURI)
		next.ServeHTTP(w, r)
		log.Debug().Msgf("Finished request to %s", r.RequestURI)
	})
}

func serverStartup(server *Server) {
	log.Info().Msgf("%s server startup: %s", server.Name, server.Addr)
	if err := server.ListenAndServe(); err != http.ErrServerClosed {
		log.Fatal().Msgf("%s server startup unsuccessful: %v", server.Name, err)
	}
}

func serverShutdown(server *Server, ctx context.Context) {
	log.Info().Msgf("%s server shutting down: %s", server.Name, server.Addr)
	if err := server.Shutdown(ctx); err != nil {
		log.Error().Msgf("%s server shutdown unsuccessful: %v", server.Name, err)
	}
	log.Info().Msgf("%s server shut down successfully", server.Name)
}

func RunServer(opts *Options) error {
	// Setup the liveness file
	path := opts.LivenessFilePath
	livenessFile, err := os.Create(path)
	if err != nil {
		log.Error().Msgf("Failed to create liveness file: %s", err)
		return fmt.Errorf("Failed to create liveness file: %s", err)
	}

	// Setup the channel for blocking
	defer os.Remove(livenessFile.Name())
	shutdown := make(chan os.Signal, 1)
	signal.Notify(shutdown, os.Interrupt, syscall.SIGTERM)
	ctx, cancel := context.WithTimeout(context.Background(), time.Second*time.Duration(opts.Wait))
	defer cancel()

	// Setup the main server
	router := mux.NewRouter()
	router.Use(loggingMiddleware)
	AddRoutes(router)
	server := &Server{
		http.Server{
			Addr:         fmt.Sprintf("%s:%d", opts.Host, opts.Port),
			WriteTimeout: time.Second * time.Duration(opts.WriteTimeout),
			ReadTimeout:  time.Second * time.Duration(opts.ReadTimeout),
			IdleTimeout:  time.Second * time.Duration(opts.IdleTimeout),
			Handler:      router,
		},
		"Main",
	}
	defer serverShutdown(server, ctx)

	// If the metrics port is not the same as the server port, run the metrics as a separate server
	if opts.Port != opts.MetricsPort {
		router := mux.NewRouter()
		router.Use(loggingMiddleware)
		AddMetricsRoute(router)
		server := &Server{
			http.Server{
				Addr:    fmt.Sprintf("%s:%d", opts.Host, opts.MetricsPort),
				Handler: router,
			},
			"Metrics",
		}

		// Run the metrics server
		go func() {
			serverStartup(server)
		}()
		defer serverShutdown(server, ctx)
		loggingMiddleware(router)
	} else {
		// Otherwise we will add the /metrics route to the main server
		AddMetricsRoute(router)
	}

	// Run the main server
	go func() {
		serverStartup(server)
	}()

	<-shutdown // Block until shutdown

	return err
}


cmd/root.go
package cmd

import (
	server "echoserver/internal"
	"fmt"
	"os"

	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

var cfgFile string

var bannerText string = `
                       _/
    _/_/      _/_/_/  _/_/_/      _/_/      _/_/_/    _/_/    _/  _/_/  _/      _/    _/_/    _/  _/_/
 _/_/_/_/  _/        _/    _/  _/    _/  _/_/      _/_/_/_/  _/_/      _/      _/  _/_/_/_/  _/_/
_/        _/        _/    _/  _/    _/      _/_/  _/        _/          _/  _/    _/        _/
 _/_/_/    _/_/_/  _/    _/    _/_/    _/_/_/      _/_/_/  _/            _/        _/_/_/  _/
`

var rootCmd = &cobra.Command{
	Use:   "echoserver",
	Short: "An HTTP echo server implementation in Go",
	Long: `An HTTP echo server implementation in Go. This will echo back to you the status
code and information you send it. This is useful for testing infrastructure.`,
	Run: func(cmd *cobra.Command, args []string) {
		banner, err := cmd.Flags().GetBool("banner")
		host, err := cmd.Flags().GetString("host")
		port, err := cmd.Flags().GetInt("port")
		metricsPort, err := cmd.Flags().GetInt("metrics-port")
		wait, err := cmd.Flags().GetInt("wait")
		writeTimeout, err := cmd.Flags().GetInt("write-timeout")
		readTimeout, err := cmd.Flags().GetInt("read-timeout")
		idleTimeout, err := cmd.Flags().GetInt("idle-timeout")
		livenessPath, err := cmd.Flags().GetString("liveness-path")
		logLevel, err := cmd.Flags().GetString("log-level")
		if err != nil {
			log.Error().Msgf("Could not get options: %s", err)
			os.Exit(1)
		}

		if logLevel == "info" {
			zerolog.SetGlobalLevel(zerolog.InfoLevel)
		} else if logLevel == "debug" {
			zerolog.SetGlobalLevel(zerolog.DebugLevel)
		} else if logLevel == "trace" {
			zerolog.SetGlobalLevel(zerolog.TraceLevel)
		}

		if banner == true {
			fmt.Println(bannerText)
		}

		opts := &server.Options{
			LivenessFilePath: livenessPath,
			Host:             host,
			Port:             port,
			MetricsPort:      metricsPort,
			Wait:             wait,
			WriteTimeout:     writeTimeout,
			ReadTimeout:      readTimeout,
			IdleTimeout:      idleTimeout,
		}
		log.Trace().Msgf("Server options: %v", opts)

		err = server.RunServer(opts)

		if err != nil {
			log.Fatal().Msg("Error shutting down server")
			os.Exit(1)
		}

		os.Exit(0)
	},
}

func Execute() {
	err := rootCmd.Execute()
	if err != nil {
		os.Exit(1)
	}
}

func init() {
	cobra.OnInitialize(initConfig)

	rootCmd.PersistentFlags().StringVar(&cfgFile, "config", "", "config file (default is $HOME/.echo.yaml)")
	rootCmd.Flags().String("log-level", "info", "Logging level. Valid choices are 'info', 'debug', and 'trace'.")
	rootCmd.Flags().String("liveness-path", "/tmp/echoserver-live", "Path to put the liveness file.")
	rootCmd.Flags().Bool("banner", true, "Print the banner on server start.")
	rootCmd.Flags().String("host", "localhost", "The host to run the server on.")
	rootCmd.Flags().Int("port", 8080, "The port to run the server on.")
	rootCmd.Flags().Int("metrics-port", 9000, "The metrics port to run the server on. This can be the same as the host port, but is set to a separate port by default for security.")
	rootCmd.Flags().Int("wait", 15, "Wait timeout in seconds.")
	rootCmd.Flags().Int("write-timeout", 15, "Write timeout in seconds.")
	rootCmd.Flags().Int("read-timeout", 15, "Read timeout in seconds.")
	rootCmd.Flags().Int("idle-timeout", 60, "Idle timeout in seconds.")

	viper.BindPFlag("banner", rootCmd.Flags().Lookup("banner"))
	viper.BindPFlag("host", rootCmd.Flags().Lookup("host"))
	viper.BindPFlag("port", rootCmd.Flags().Lookup("port"))
	viper.BindPFlag("metrics-port", rootCmd.Flags().Lookup("metrics-port"))
	viper.BindPFlag("wait", rootCmd.Flags().Lookup("wait"))
	viper.BindPFlag("write-timeout", rootCmd.Flags().Lookup("write-timeout"))
	viper.BindPFlag("read-timeout", rootCmd.Flags().Lookup("read-timeout"))
	viper.BindPFlag("idle-timeout", rootCmd.Flags().Lookup("idle-timeout"))
}

func initConfig() {
	if cfgFile != "" {
		viper.SetConfigFile(cfgFile)
	} else {
		home, err := os.UserHomeDir()
		cobra.CheckErr(err)

		viper.AddConfigPath(home)
		viper.SetConfigType("yaml")
		viper.SetConfigName(".echoserver")
	}

	viper.AutomaticEnv()

	if err := viper.ReadInConfig(); err == nil {
		fmt.Fprintln(os.Stderr, "Using config file:", viper.ConfigFileUsed())
	}
}
